logger.go

```diff
@@ -28,33 +28,346 @@ import (
+	var str string
+	if value != nil {
+		str = value.String()
+	}
+	return newLogField(name, logFieldValueString(str))
+}
+
+func newLogFieldError(name string, value error) LogField {
+	var str string
+	if value != nil {
+		str = value.Error()
+	}
+	return newLogField(name, logFieldValueString(str))
+}
+
+func newLogFieldStringer(name string, value fmt.Stringer) LogField {
+	var str string
+	if value != nil {
+		str = value.String()
+	}
+	return newLogField(name, logFieldValueString(str))
+}
+
+func newLogFieldString(name string, value string) LogField {
+	return newLogField(name, logFieldValueString(value))
+}
+
+func newLogFieldInt(name string, value int) LogField {
+	return newLogField(name, logFieldValueInt64(int64(value)))
+}
+
+func newLogFieldBool(name string, value bool) LogField {
+	return newLogField(name, logFieldValueBool(value))
+}
+
+type StructuredLogger interface {
+	Error(msg string, fields ...LogField)
+	Warning(msg string, fields ...LogField)
+	Info(msg string, fields ...LogField)
+	Debug(msg string, fields ...LogField)
+}
+
+// A LogFieldValue can represent any Go value, but unlike type any,
+// it can represent most small values without an allocation.
+// The zero Value corresponds to nil.
+type LogFieldValue struct {
+	num uint64
+	any interface{}
+}
+
+// LogFieldValueType is the type of a [LogFieldValue].
+type LogFieldValueType int
+
+// It's important that LogFieldTypeAny is 0 so that a zero Value represents nil.
+const (
+	LogFieldTypeAny LogFieldValueType = iota
+	LogFieldTypeBool
+	LogFieldTypeInt64
+	LogFieldTypeString
+)
+
+// LogFieldValueType returns v's LogFieldValueType.
+func (v LogFieldValue) LogFieldValueType() LogFieldValueType {
+	switch x := v.any.(type) {
+	case LogFieldValueType:
+		return x
+	case string:
+		return LogFieldTypeString
+	default:
+		return LogFieldTypeAny
+	}
+}
+
+func logFieldValueString(value string) LogFieldValue {
+	return LogFieldValue{num: uint64(len(value)), any: value}
+}
 
-type defaultLogger struct{}
+func logFieldValueInt(v int) LogFieldValue {
+	return logFieldValueInt64(int64(v))
+}
+
+func logFieldValueInt64(v int64) LogFieldValue {
+	return LogFieldValue{num: uint64(v), any: LogFieldTypeInt64}
+}
+
+func logFieldValueBool(v bool) LogFieldValue {
+	u := uint64(0)
+	if v {
+		u = 1
+	}
+	return LogFieldValue{num: u, any: LogFieldTypeBool}
+}
+
+// Any returns v's value as an interface.
+func (v LogFieldValue) Any() interface{} {
+	switch v.LogFieldValueType() {
+	case LogFieldTypeAny:
+		if k, ok := v.any.(LogFieldValueType); ok {
+			return k
+		}
+		return v.any
+	case LogFieldTypeInt64:
+		return int64(v.num)
+	case LogFieldTypeString:
+		return v.str()
+	case LogFieldTypeBool:
+		return v.bool()
+	default:
+		panic(fmt.Sprintf("bad value type: %s", v.LogFieldValueType()))
+	}
+}
+
+// String returns Value's value as a string, formatted like [fmt.Sprint]. Unlike
+// the methods Int64, Float64, and so on, which panic if v is of the
+// wrong kind, String never panics.
```

Comment:
It does panic if the type is unknown. Not sure if that should change or if we should update this documentation.